[client] Skip open files on cache cleanup when possible

christge-sft opened the pull request:
On the Algorithmic side open files are handled as if they were pinned.

The tricky part is how the PosixQuotaManager (PQM) will get to know which files are open. In this approach, open hashes are communicated to the PQM using a dedicated magic xattr.

Firstly the PQM needs to know its managed mount points. This is happening early, when the PQM is acquired.
To get the open hashes of a mount point, this approach introduces the magic xattr `user.list_open_hashes`.
To get the mount points managed from the shared PQM,  this approach introduces the magic xattr `user.list_managed_mountopints`.
To get the open hashes of all the mount points associated to the shared PQM, this approach introduces the magic xattr `user.list_groups_open_hashes`. This last magic xattr uses exactly the same mechanism that is employed in `DoCleanup()`.


By default on this stage the feature is turned off. To enable it add `CVMFS_CACHE_CLEANUP_NONOPENLRU=yes` in the configuration.

christge-sft committed:
[CORE] create backend for open file aware LRU $ cleanup

christge-sft committed:
[INTEGRATION] add PosixCacheSettings option for this new policy

christge-sft committed:
[INTEGRATION] make the feature runtime selectible

christge-sft committed:
[INTEGRATION] check if feature is enabled via an xattr

christge-sft committed:
[TESTING] add helpers for testing

christge-sft committed:
[INTEGRATION] list open hashes via a magic xattr per repo

christge-sft committed:
[INTEGRATION] allow QM to keep track of the managed mountpoints

christge-sft committed:
[INTEGRATION] create a magic xattr to rertrieve the associated mountpoints

christge-sft committed:
fixup! [INTEGRATION] allow QM to keep track of the managed mountpoints

christge-sft committed:
fixup! [INTEGRATION] allow QM to keep track of the managed mountpoints

christge-sft committed:
[INTEGRATION] propagate the cleanup policy to the Quota Manager based on settings parsing

christge-sft committed:
[CORE] prepare to the core of the cleanup policy. Last missing part collect the hashes from every mountpoint via the magic xattr list_open_hashes

christge-sft committed:
Revert "[CORE] prepare to the core of the cleanup policy. Last missing part collect the hashes from every mountpoint via the magic xattr list_open_hashes"

This reverts commit 04375c55efe9e3a98e4fbb4476ff18534809a147.

christge-sft committed:
[INTEGRATION] create a mechanism to retrieve the open hashes of the group as a magic xattr

christge-sft committed:
[INTEGRATION] spawn open hashes collectors from the main QM thread

christge-sft committed:
[INTEGRATION] establish a timeout mechanism for hash collection

christge-sft committed:
[INTEGRATION] Collect hashes from a mountpoint

christge-sft committed:
fixup! fixup! [INTEGRATION] allow QM to keep track of the managed mountpoints

christge-sft committed:
fixup! [INTEGRATION] spawn open hashes collectors from the main QM thread

christge-sft committed:
[TESTING] check that the added magic xattrs are behaving properly

christge-sft committed:
[TESTING] compare the two cleanup strategies

christge-sft committed:
[TESTING] return on error a different value

christge-sft committed:
adopt clang-tidy const suggestions

christge-sft committed:
fixup! [CORE] prepare to the core of the cleanup policy. Last missing part collect the hashes from every mountpoint via the magic xattr list_open_hashes

christge-sft committed:
[TESTING] Use unit testing for the vanilla version; To test this feature it is better to use integration testing

christge-sft committed:
fixup! fixup! [CORE] prepare to the core of the cleanup policy. Last missing part collect the hashes from every mountpoint via the magic xattr list_open_hashes

christge-sft committed:
[INTEGRATION] restrict the use of collect groups hashes when cleanup_unused_first is disabled

christge-sft committed:
[TESTING] test for (groups)open hashes and managed mountpoints only when cleanup_unused_first is enabled

christge-sft committed:
remove unsupported CPP features

christge-sft committed:
adjust to MACOSX requirements. Feature not supported for MACOS

christge-sft committed:
fixup! adjust to MACOSX requirements. Feature not supported for MACOS

christge-sft committed:
fixup! fixup! adjust to MACOSX requirements. Feature not supported for MACOS

vvolkl commented on cvmfs/cache_posix.cc:
```diff
@@ -269,9 +269,9 @@ PosixCacheManager *PosixCacheManager::Create(
+    const bool do_refcount, const bool cleanup_unused_first) {
   UniquePtr<PosixCacheManager> cache_manager(
-      new PosixCacheManager(cache_path, alien_cache, do_refcount));
+      new PosixCacheManager(cache_path, alien_cache, do_refcount,cleanup_unused_first));
```
```suggestion
      new PosixCacheManager(cache_path, alien_cache, do_refcount, cleanup_unused_first));
```
Whitespace. Can you rerun clang-fmt on changed files?

christge-sft committed:
apply clang-format on changed files

vvolkl commented on cvmfs/magic_xattr.cc:
```diff
@@ -8,11 +8,13 @@
 #include <string>
 #include <vector>
 
+#include "cache_posix.h"  // PosixCacheManager
```
This type of comment is likely to go stale quickly, when another PR uses something from the header but does not update the comment, let's remove these.

vvolkl commented on cvmfs/mountpoint.cc:
```diff
@@ -306,6 +306,12 @@ FileSystem::PosixCacheSettings FileSystem::DeterminePosixCacheSettings(
+  if (options_mgr_->GetValue(
+          MkCacheParm("CVMFS_CACHE_CLEANUP_NONOPENLRU", instance), &optarg)
+      && options_mgr_->IsOn(optarg)) {
+    settings.cleanup_unused_first = settings.do_refcount;
```
Add a comment to explain that nonopenlru is not possible without the refcounted cache manager

vvolkl commented on cvmfs/mountpoint.cc:
```diff
@@ -682,7 +688,7 @@ CacheManager *FileSystem::SetupPosixCacheMgr(const string &instance) {
       settings.avoid_rename ? PosixCacheManager::kRenameLink
                             : PosixCacheManager::kRenameNormal,
-      settings.do_refcount));
+      settings.do_refcount,settings.cleanup_unused_first));
```
```suggestion
      settings.do_refcount, settings.cleanup_unused_first));
```
whitespace

vvolkl commented on cvmfs/quota_posix.cc:
```diff
@@ -27,6 +27,9 @@
 #include <sys/stat.h>
+#include <sys/xattr.h>
+
+#include <algorithm>  //std::all_of
```
Remove comment (see above)

vvolkl commented on cvmfs/quota_posix.cc:
```diff
@@ -35,6 +38,7 @@
 #include <sys/wait.h>
 #include <unistd.h>
 
+#include <algorithm>
```
This include is duplicated, right?

vvolkl commented on cvmfs/quota_posix.cc:
```diff
@@ -2106,3 +2256,91 @@ void PosixQuotaManager::ManagedReadHalfPipe(int fd, void *buf, size_t nbyte) {
+        joined[i] = true;
+      }
+    }
+    i = (++i) % handlers.size();
```
There's a compiler warning here:
```
warning: operation on ‘i’ may be undefined [-Wsequence-point]
```
Apparently the order of ++ and % is ambiguous here.

vvolkl commented on cvmfs/magic_xattr.cc:
```diff
@@ -72,6 +74,13 @@ MagicXattrManager::MagicXattrManager(
+
+  Register("user.cleanup_unused_first", new CleanupUnusedFirstMagicXattr());
+  Register("user.list_open_hashes", new ListOpenHashesMagicXattr());
+  Register("user.list_groups_open_hashes",
```
`"group" `is a slightly confusing name in this context. I would replace it with `"all_repositories"`

vvolkl replied:
This is useful for debugging, but it doesn't need to be an xattr. I would suggest to make it a cvmfs_talk command instead, something like:

```
cvmfs_talk -i <repo> cache openhashes
```

vvolkl commented on cvmfs/magic_xattr.cc:
```diff
@@ -72,6 +74,13 @@ MagicXattrManager::MagicXattrManager(
+  Register("user.list_open_hashes", new ListOpenHashesMagicXattr());
+  Register("user.list_groups_open_hashes",
+           new ListGroupsOpenHashesMagicXattr());
+  Register("user.list_managed_mountpoints",
```
Same here, having this as an xattr for a specific repository is a bit unexpected, I would also make this a cvmfs_talk command

vvolkl reviewed (changes requested):
Very nice! Some comments inline.

As discussed, a future optimization could be to only exchange partial hashes, since a collision has negligible downsides. Can be done in a future PR together with the std::set optimization.

christge-sft committed:
Remove magic xattrs that are not directly associated with the mountpoint.

Extended attributes list_managed_mountpoints and list_groups_open_hashes are not directly associated with the mountpoint. They are associated with the shared quota manager. If the information is needed, it's better to be obtained using cvmfs_talk.
Signed-off-by: Georgios Christodoulis <christge@cern.ch>

christge-sft committed:
cleanup headers

christge-sft committed:
Readability - maintenance

christge-sft committed:
resolve ambiguity in operators priority

vvolkl commented on cvmfs/quota_posix.h:
```diff
@@ -372,6 +385,29 @@ class PosixQuotaManager : public QuotaManager {
+  bool cleanup_unused_first_;
+  std::vector<std::string> mountpoints_;
+
+  std::vector<shash::Any> CollectGroupsHashes();
```
I still don't like this name. What about 
```suggestion
  std::vector<shash::Any> CollectAllOpenHashes();
```
?

vvolkl reviewed (changes requested):
See comment inline

christge-sft committed:
Remove the "Group" notation. Fix mutex leak

"Group" can mean many things in many contexts, what we wanted here is
the notion of all the open hashes managed by this particular Shared Quota
Manager.

The mutex free should happen regardless of wether the PosixQuotaManager
is initialized or not, since it's allocated during construction.

Signed-off-by: Georgios Christodoulis <christge@cern.ch>

vvolkl reviewed (approved):
LGTM!
